nomad/core_sched.go

```diff
@@ -892,6 +895,94 @@ func (c *CoreScheduler) rootKeyRotation(eval *structs.Evaluation) (bool, error)
 	return true, nil
 }
 
+// secureVariablesReKey is optionally run after rotating the active
+// root key. It iterates over all the variables for the non-active
+// keys, decrypts them, and re-encrypts them in batches with the
+// currently active key. This job does not GC the keys, which is
+// handled in the normal periodic GC job.
+func (c *CoreScheduler) secureVariablesRekey(eval *structs.Evaluation) error {
+
+	ws := memdb.NewWatchSet()
+	iter, err := c.snap.RootKeyMetas(ws)
+	if err != nil {
+		return err
+	}
+
+	for {
+		raw := iter.Next()
+		if raw == nil {
+			break
+		}
+		keyMeta := raw.(*structs.RootKeyMeta)
+		if keyMeta.Active {
+			continue // never rotate
+		}
+		varIter, err := c.snap.GetSecureVariablesByKeyID(ws, keyMeta.KeyID)
+		if err != nil {
+			return err
+		}
+		err = c.batchRotateVariables(varIter, eval)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// rootKeyFullRotatePerKey runs over an iterator of secure variables
+// and decrypts them, and then sends them back as batches to be
+// re-encrypted with the currently active key.
+func (c *CoreScheduler) batchRotateVariables(iter memdb.ResultIterator, eval *structs.Evaluation) error {
+
+	upsertFn := func(variables []*structs.SecureVariableDecrypted) error {
+		if len(variables) == 0 {
+			return nil
+		}
+		args := &structs.SecureVariablesUpsertRequest{
+			Data: variables,
+			WriteRequest: structs.WriteRequest{
+				Region:    c.srv.config.Region,
+				AuthToken: eval.LeaderACL, // TODO: will this work?
+			},
+		}
+		reply := &structs.SecureVariablesUpsertResponse{}
+		return c.srv.RPC("SecureVariables.Upsert", args, reply)
```

Comment:
Using Upsert will change the ModifyIndex while rekeying the variable. This will cause CAS conflicts on items that haven't haven't been modified from the user's perspective. How do we feel about that?

Reply:
I _think_ that's actually what we want. If this job is running you probably _don't_ want to be able to accidentally overwrite anything until it's done with that batch. I don't love that it ends up resolving any open blocking queries though.